java/core/src/test/java/com/lancedb/lance/VectorSearchTest.java

```diff
@@ -0,0 +1,240 @@
+ * limitations under the License.
+ */
+
+package com.lancedb.lance;
+
+import org.apache.arrow.dataset.scanner.Scanner;
+import org.apache.arrow.vector.Float4Vector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import com.lancedb.lance.ipc.Query;
+import com.lancedb.lance.ipc.ScanOptions;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+// Creates a dataset with 5 batches where each batch has 80 rows
+//
+// The dataset has the following columns:
+//
+//  i   - i32      : [0, 1, ..., 399]
+//  s   - &str     : ["s-0", "s-1", ..., "s-399"]
+//  vec - [f32; 32]: [[0, 1, ... 31], [32, ..., 63], ... [..., (80 * 5 * 32) - 1]]
+//
+// An IVF-PQ index with 2 partitions is trained on this data
+public class VectorSearchTest {
+  @TempDir
+  Path tempDir;
+
+  @Test
+  void test_create_index() throws Exception {
+    try (TestVectorDataset testVectorDataset = new TestVectorDataset(tempDir.resolve("test_create_index"))) {
+      try (Dataset dataset = testVectorDataset.create()) {
+        testVectorDataset.createIndex(dataset);
+        List<String> indexes = dataset.listIndexes();
+        assertEquals(1, indexes.size());
+        assertEquals(TestVectorDataset.indexName, indexes.get(0));
+      }
+    }
+  }
+
+  // rust/lance-linalg/src/distance/l2.rs:256:5:
+  // 5assertion `left == right` failed
+  // Directly panic instead of throwing an exception
+  // @Test
+  // void search_invalid_vector() throws Exception {
+  //   try (TestVectorDataset testVectorDataset = new TestVectorDataset(tempDir.resolve("test_create_index"))) {
+  //     try (Dataset dataset = testVectorDataset.create()) {
+  //       float[] key = new float[30];
+  //       for (int i = 0; i < 30; i++) {
+  //         key[i] = (float) (i + 30);
+  //       }
+  //       ScanOptions options = new ScanOptions.Builder()
+  //           .nearest(new Query.Builder()
+  //               .setColumn(TestVectorDataset.vectorColumnName)
+  //               .setKey(key)
+  //               .setK(5)
+  //               .setUseIndex(false)
+  //               .build())
+  //           .build();
+  //       assertThrows(IllegalArgumentException.class, () -> {
+  //         try (Scanner scanner = dataset.newScan(options)) {
+  //           try (ArrowReader reader = scanner.scanBatches()) {
+  //           }
+  //         }
+  //       });
+  //     }
+  //   }
+  // }
+
+  @ParameterizedTest
+  @ValueSource(booleans = { false, true })
+  void test_knn(boolean createVectorIndex) throws Exception {
```

Comment:
@walterddr this is the test similar to the doc one,
which tested KNN and ANN(if create vector index).

The website one is more high level Python API.
Java API mocks the low-level Rust API:
 using the Scanner(ScanOptions.nearest(Query.setKey(the nearest key)))